pds4indextools/pds4_create_xml_index.py

```diff
@@ -25,45 +31,118 @@
+    --simplify-xpaths    Replace unique XPath segments with shortened versions.
     --output-file OUTPUT_FILE
                          The output path and filename for the resulting index file.
     --verbose            Activate verbose printed statements during runtime.
     --sort-by SORT_BY    Sort the index file by a chosen set of columns.
     --clean-header-field-names
                          Replace the ":" and "/" with Windows-friendly characters.
-    --extra-file-info EXTRA_FILE_INFO  
+    --extra-file-info EXTRA_FILE_INFO
                          Add additional column(s) to the index file containing file or
                          bundle information. Possible values are: "LID", "filename",
                          "filepath", "bundle", and "bundle_lid". Multiple values may be
                          specified separated by spaces.
     --config-file CONFIG_FILE
                          An optional .ini configuration file for further customization.
+    --dump-available-xpaths
+                         Create a .txt file containing all available XPath headers for
+                         given label file(s). Can be modified and used as a file for
+                         --elements-file
+    --fixed-width        Create a fixed-width index file. 
 
 Example:
-python3 pds4_create_xml_index.py <toplevel_directory> "glob_path1" "glob_path2" 
---output_file <outputfile> --elements-file sample_elements.txt --verbose
+    python3 pds4_create_xml_index.py <toplevel_directory> "glob_path1" "glob_path2"
+    --output_file <outputfile> --elements-file sample_elements.txt --verbose
 """
 
 import argparse
+from collections import namedtuple
 import configparser
+import fnmatch
+import functools
+from itertools import groupby
 from lxml import etree
 import pandas as pd
 from pathlib import Path
 import requests
 import sys
 
 
+SplitXPath = namedtuple('SplitXPath',
+                        ['xpath', 'parent', 'child', 'prefix', 'num'])
+
+
+def convert_header_to_xpath(root, xpath_find, namespaces):
+    """
+    Replace hierarchical components of XPath with attribute names and namespaces.
+
+    While the XPaths are accurate to the hierarchy of the elements referenced, they
+    provide no information on their own without the attributed label file for reference.
+    This function replaces the asterisks with the respective names of the elements and
+    attributes they represent.
+
+    Args:
+        root (Element): The root element of the XML document.
+        xpath_find (str): Original XML header path.
+        namespaces (dict): Dictionary of XML namespace mappings.
+
+    Returns:
+        str: Converted XPath expression.
+    """
+    sections = xpath_find.split('/')
+    xpath_final = ''
+    portion = ''
+    for sec in sections[1:]:
+        portion = portion + '/' + sec
+        tag = str(root.xpath(portion, namespaces=namespaces)[0].tag)
+        if '*' in sec:
```

Comment:
If the unprocessed XPath headers always followed the same syntax, then this would apply. However, this isn't the case. Below are a collection of XPath headers when they are first scraped with `lxml`:

```
/*/*[2]/*[6]/*[2]
/*/*[2]/*[6]/*[3]/*[1]
/*/*[2]/*[6]/*[3]/*[2]
/*/*[2]/*[7]/cassini:Cassini/cassini:ISS_Specific_Attributes/cassini:mission_phase_name
/*/*[2]/*[7]/cassini:Cassini/cassini:ISS_Specific_Attributes/cassini:spacecraft_clock_count_partition
/*/*[2]/*[7]/cassini:Cassini/cassini:ISS_Specific_Attributes/cassini:spacecraft_clock_start_count
```

depending on the namespace, `lxml` will give some XPaths different formats. The ones that are just asterisks and predicates are usually the default namespace (`pds`), while the half-and-half syntax is for any other namespace. 

In short, `sec.startswith('*') ` can apply here, but `sec.startswith(':')` doesn't. It needs to remain `if ':' in sec:`

Reply:
Got it. It was mainly the first one I was concerned about. It seemed like a potential bug that you look for a * anywhere, but then delete the first character even if it's potentially not a *. Using `startswith` guarantees that you're always deleting a *.